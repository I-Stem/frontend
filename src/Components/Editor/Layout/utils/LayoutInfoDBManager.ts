import { getObject, putObject, saveObject } from "@utils/indexedDb";
import { Editor } from "../../EditorTypings";
import { BoundingBoxProperty } from "../CanvasTypes";
const dbName = "i-stem";

export type LayoutEditorBoundingBoxInfo = {
  [key in string]: BoundingBoxProperty[];
};

export const getLayoutEditorBoundingBoxForPage = async (
  fileId: string,
  currentPage: number
) => {
  return new Promise((resolve, reject) => {
    getObject(
      dbName,
      "layout",
      `${fileId}-box`,
      (data: LayoutEditorBoundingBoxInfo) => {
        if (!data || !data[currentPage]) {
          return resolve(null);
        }

        return resolve(data[currentPage]);
      }
    );
  });
};

/*
export const getLayoutAndTextInfoForPage = async (fileId: string, currentPage:number):Promise<{[key in string]:PageLayoutAndTextInfo} | null> => {
    return new Promise((resolve, reject) => {
        getObject("i-stem", "layout", `${fileId}-layout`, async (data: PagewiseLayoutAndTextInfo) => {
            if(data && currentPage in data)
            resolve(data[currentPage]);
            else
            resolve(null);
    });
});
}
*/

export const saveBoundingBoxesInIndexedDB = (
  fileId: string,
  currentPage: number,
  boundingBoxes: Array<BoundingBoxProperty>
) => {
  const keyPath = `${fileId}-box`;
  getObject(
    "i-stem",
    "layout",
    keyPath,
    (data: { [pageNumber: string]: BoundingBoxProperty[] }) => {
      if (data !== null) {
        putObject(
          dbName,
          "layout",
          { ...data, [currentPage]: boundingBoxes },
          "id",
          keyPath
        );
      } else
        saveObject(
          dbName,
          "layout",
          "id",
          { [currentPage]: boundingBoxes },
          keyPath
        );
    }
  );
};

export interface IndexedDBBoundingBoxSchema {
  //1 based page numbers
  [pageNumber: number]: BoundingBoxProperty[];
}
export const getBoundingBoxesForPageFromIndexedDb = async (
  fileId: string,
  currentPage: number
): Promise<BoundingBoxProperty[] | null> => {
  return new Promise((resolve, reject) => {
    getObject(
      dbName,
      "layout",
      `${fileId}-box`,
      (data: IndexedDBBoundingBoxSchema) => {
        if (data === null || !(currentPage in data)) {
          return resolve(null);
        }

        return resolve(data[currentPage]);
      }
    );
  });
};

export const updateCanvasBoxesWithReadingOrder = async (
  fileId: string,
  currentPage: number,
  boxes: BoundingBoxProperty[],
  layoutWithText: Editor.BoundingBoxWithCorrectedOCRText[]
) => {
  for (let i = 0; i < layoutWithText.length; i++) {
    boxes.filter(box => box.ind === layoutWithText[i].id)[0].order =
      layoutWithText[i].reading_order;
  }

  await saveBoundingBoxesInIndexedDB(fileId, currentPage, boxes);

  return boxes;
};

export const updateExistingBoundingBoxes = async (
  fileId: string,
  currentPage: number,
  updatedBoxes: BoundingBoxProperty[]
) => {
  const boxes =
    (await getBoundingBoxesForPageFromIndexedDb(fileId, currentPage)) || [];
  const result = updatedBoxes.forEach(box => {
    const boxIndex = boxes.findIndex(oldBox => box.ind === oldBox.ind);
    boxes[boxIndex] = box;
  });

  saveBoundingBoxesInIndexedDB(fileId, currentPage, boxes);
};
