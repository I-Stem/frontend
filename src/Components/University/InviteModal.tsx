import { GreenButton } from "@Components/HOC/Dashboard";
import { Formik } from "formik";
import React from "react";
import { Form, Modal } from "react-bootstrap";
import * as Yup from "yup";

export const InviteModal: React.FC<Props> = props => {
  const { isOpen, toggleModal, children, formSubmit, invitationFor } = props;
  const initialValues = {
    emails: "",
  };
  const handleSubmit = (values: any) => {
    formSubmit(values);
  };
  return (
    <Modal show={isOpen} onHide={toggleModal} animation>
      <Modal.Header closeButton>
        <Modal.Title>
          <h3 className="lip-title">INVITE {invitationFor}</h3>
        </Modal.Title>
      </Modal.Header>
      <Formik
        initialValues={initialValues}
        onSubmit={handleSubmit}
        validationSchema={Yup.object().shape({
          emails: Yup.string().required("Emails cannot be blank"),
        })}
      >
        {formik => (
          <Form onSubmit={formik.handleSubmit}>
            <Modal.Body>
              {children}
              <Form.Group controlId="emails">
                <Form.Control
                  className="stud-search-box email"
                  placeholder={`Enter ${invitationFor.toLowerCase()} email`}
                  {...formik.getFieldProps("emails")}
                />
              </Form.Group>
              {formik.errors.emails && formik.touched.emails ? (
                <div className="error">{formik.errors.emails}</div>
              ) : null}
            </Modal.Body>
            <Modal.Footer>
              <div style={{ width: "40%" }}>
                <GreenButton htmlType="submit">
                  <span className="flex items-center">
                    <span className="ml-2">SEND INVITE</span>
                  </span>
                </GreenButton>
              </div>
            </Modal.Footer>
          </Form>
        )}
      </Formik>
    </Modal>
  );
};

interface Props {
  isOpen: boolean;
  toggleModal: Function;
  formSubmit: (data: any) => void;
  invitationFor: string;
}
