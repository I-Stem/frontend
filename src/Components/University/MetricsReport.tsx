import React, { useEffect, useState } from "react";
import { Modal } from "react-bootstrap";
import { GreenButton } from "@Components/HOC/Dashboard";
import { UniversityPortal } from "@Services";

export const MetricsReportDialog: React.FC<MetricsReportDialogProps> = (
  props: MetricsReportDialogProps
) => {
  const { showModal, toggleDialog } = props;
  const [open, setOpen] = useState(showModal);
  useEffect(() => {
    setOpen(showModal);
  }, [showModal]);
  const handleClose = () => {
    toggleDialog();
    UniversityPortal.emailStudentsReport();
  };
  return (
    <Modal show={open} onHide={toggleDialog} animation>
      <Modal.Header closeButton>
        <Modal.Title>Generate Report</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        A report for students metrics will be generated as CSV and send to your
        email.
      </Modal.Body>
      <Modal.Footer>
        <div style={{ width: "40%" }}>
          <GreenButton htmlType="button" onClick={handleClose}>
            <span className="flex items-center">
              <span className="ml-2">Send Email</span>
            </span>
          </GreenButton>
        </div>
      </Modal.Footer>
    </Modal>
  );
};

interface MetricsReportDialogProps {
  showModal: boolean;
  toggleDialog: () => void;
}
