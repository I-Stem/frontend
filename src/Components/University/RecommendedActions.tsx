import { BlueButton } from "@Components/HOC/Dashboard";
import React, { useState } from "react";
import { Col, Row } from "react-bootstrap";
import "./style.scss";
import fileNames from "@Definitions/Constants/image";
import { STUDENTS, SETTINGS } from "@Definitions/Constants/universityRoutes";
import { UniversityPortal } from "@Services";
import { CardPreferences } from "@Interfaces";
import { UserType } from "@Definitions/Constants";

const { ONBOARD_STUDENTS, ONBOARD_STAFF, CLOSE_ICON } = fileNames;

export const RecommendedActions = (props: Props) => {
  const { cardPreferences, updatePreferences, userType } = props;
  const [staffCard, toggleStaffCard] = useState(
    cardPreferences?.showOnboardStaffCard
  );
  const [studentCard, toggleStudentCard] = useState(
    cardPreferences?.showOnboardStudentsCard
  );
  const closeTile = (cardType: string) => {
    if (cardType === "students") {
      toggleStudentCard(!studentCard);
      UniversityPortal.updateUserCardPreferences({
        showOnboardStudentsCard: false,
        showOnboardStaffCard: staffCard,
      });
      updatePreferences({
        showOnboardStaffCard: staffCard,
        showOnboardStudentsCard: false,
      });
    }
    if (cardType === "staffs") {
      toggleStaffCard(!staffCard);
      UniversityPortal.updateUserCardPreferences({
        showOnboardStaffCard: false,
        showOnboardStudentsCard: studentCard,
      });
      updatePreferences({
        showOnboardStaffCard: false,
        showOnboardStudentsCard: studentCard,
      });
    }
  };
  const onboardingUser = () => {
    if (userType === UserType.UNIVERSITY) {
      return "STUDENTS";
    }
    if (userType === UserType.BUSINESS) {
      return "EMPLOYEES";
    }
  };

  return (
    <div>
      {staffCard === true || studentCard === true ? (
        <h2 className="font-semibold text-xl heading-color">
          RECOMMENDED ACTIONS
        </h2>
      ) : (
        <></>
      )}

      {studentCard && (
        <Row className="onboard-tiles">
          <Col style={{ marginTop: "1rem" }} sm={6}>
            <h3 className="font-semibold text-l heading-color">
              Get started by inviting your{" "}
              {onboardingUser()?.toLocaleLowerCase()} to the I-Stem platform.
            </h3>
            <p>
              Or you can always go to{" "}
              {onboardingUser()
                ?.charAt(0)
                .concat(
                  onboardingUser()!
                    .substring(1)
                    .toLowerCase()
                )}{" "}
              &gt; INVITE {onboardingUser()} / IMPORT {onboardingUser()} to
              invite {onboardingUser()?.toLocaleLowerCase()}.
            </p>
            <div className="onboard-button">
              <BlueButton href={STUDENTS}>
                ONBOARD {onboardingUser()}
              </BlueButton>
            </div>
          </Col>
          <Col style={{ marginLeft: "8rem" }} sm={4}>
            <img
              className="w-full lip-img"
              src={ONBOARD_STUDENTS}
              alt="decorative"
              aria-hidden="true"
            />
          </Col>
          <div onClick={event => closeTile("students")}>
            <img
              className="close-tiles"
              src={CLOSE_ICON}
              alt="Close tile"
              role="button"
            />
          </div>
        </Row>
      )}
      {staffCard && (
        <Row className="onboard-tiles">
          <Col style={{ marginTop: "1rem" }} sm={6}>
            <h3 className="font-semibold text-l heading-color">
              Get started by inviting your staff to the I-Stem platform.
            </h3>
            <p>
              Or you can always go to Settings &gt; INVITE STAFF to invite
              staff.
            </p>
            <div className="onboard-button">
              <BlueButton href={SETTINGS}>ONBOARD STAFF</BlueButton>
            </div>
          </Col>
          <Col style={{ marginLeft: "8rem" }} sm={4}>
            <img
              className="w-full lip-img"
              src={ONBOARD_STAFF}
              alt="decorative"
              aria-hidden="true"
            />
          </Col>
          <div onClick={event => closeTile("staffs")}>
            <img
              className="close-tiles"
              src={CLOSE_ICON}
              alt="Close tile"
              role="button"
            />
          </div>
        </Row>
      )}
    </div>
  );
};

interface Props {
  cardPreferences: CardPreferences;
  updatePreferences: (preferences: CardPreferences) => void;
  userType: UserType;
}
