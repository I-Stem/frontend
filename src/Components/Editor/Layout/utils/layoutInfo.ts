import { Editor, State } from "@Components/Editor/EditorTypings";
import { PageActivityInterface } from "@Interfaces";
import { EditorServices } from "@Services/API/Editor";
import { getObject, putObject, saveObject } from "@utils/indexedDb";
import { RefObject } from "react";
import { BoundingBoxProperty } from "../CanvasTypes";
import { toJson } from "./canvasUtils";
import { mapCanvasBoundingBoxesToRemediationAPISchema } from "./parseData";

/*
export interface PageLayoutAndTextInfo {
  layoutInfo: Editor.BoundingBoxWithCorrectedOCRText;
}

export interface PagewiseLayoutAndTextInfo {
  [key: string]: {
    [key in State]: PageLayoutAndTextInfo;
  };
}

const versionMaintainerForLayoutInfo = (
  data: PagewiseLayoutAndTextInfo,
  currentPage: number,
  layoutInfo: Editor.BoundingBoxWithCorrectedOCRText[]
) => {
  let pageData = {};
  if (currentPage in data) {
    const currentPageData = data[currentPage];
    if (1 in currentPageData) {
      //delete currentPageData[0];
      pageData = {
        0: { layoutInfo: currentPageData[1].layoutInfo },
        1: { layoutInfo },
      };
    } else {
      pageData = {
        ...currentPageData,
        1: { layoutInfo },
      };
    }
  } else {
    pageData = {
      0: {
        layoutInfo,
      },
    };
  }

  return pageData;
};
*/

export const getRelativeBoxesInX1Y1X2Y2FromRelativeXYWH = (
  relativeXYWH: Editor.BoundingBoxType
): Editor.BoundingBoxType => {
  return [
    relativeXYWH[0],
    relativeXYWH[1],
    relativeXYWH[0] + relativeXYWH[2],
    relativeXYWH[1] + relativeXYWH[3],
  ];
};

export const getRelativeBoxInXYWHFromX1Y1X2Y2 = (
  relativeX1Y1X2Y2: Editor.BoundingBoxType
): Editor.BoundingBoxType => {
  return [
    relativeX1Y1X2Y2[0],
    relativeX1Y1X2Y2[1],
    relativeX1Y1X2Y2[2] - relativeX1Y1X2Y2[0],
    relativeX1Y1X2Y2[3] - relativeX1Y1X2Y2[1],
  ];
};

export const getScaledUpBox = (
  relativeBox: Editor.BoundingBoxType,
  canvasWidth: number,
  canvasHeight: number
): Editor.BoundingBoxType => {
  return [
    relativeBox[0] * canvasWidth,
    relativeBox[1] * canvasHeight,
    relativeBox[2] * canvasWidth,
    relativeBox[3] * canvasHeight,
  ];
};

export const getScaledDownBox = (
  absoluteBox: Editor.BoundingBoxType,
  canvasWidth: number,
  canvasHeight: number
) => {
  return [
    absoluteBox[0] / canvasWidth,
    absoluteBox[1] / canvasHeight,
    absoluteBox[2] / canvasWidth,
    absoluteBox[3] / canvasHeight,
  ];
};

export const getTextInfoForLayoutFromRemediationAPI = async (
  fileId: string,
  currentPage: number,
  boundingBoxes: Array<BoundingBoxProperty>,
  twoColumn: boolean,
  analytics: PageActivityInterface,
  type: string,
  version: string
): Promise<{
  hasError: boolean;
  errorMessage: string | null;
  data: Editor.BoundingBoxWithCorrectedOCRText[];
}> =>
  new Promise((resolve, reject) => {
    toJson(`${fileId}-json`, async (json: Editor.OCRJsonSchema) => {
      const layoutInfo = mapCanvasBoundingBoxesToRemediationAPISchema(
        boundingBoxes,
        currentPage - 1,
        json,
        type,
        version,
        twoColumn
      );
      if (!layoutInfo) {
        return resolve({
          hasError: true,
          errorMessage:
            "Couldn't make request for text, error while data conversion",
          data: [],
        });
      }

      try {
        const res = await EditorServices.layoutText(currentPage, fileId, {
          analytics,
          recreationApi: layoutInfo,
        });

        console.log("api response: " + JSON.stringify(res));
        if (!res) {
          return resolve({
            hasError: true,
            errorMessage:
              "Couldn't get text data from backend, please try again.",
            data: [],
          });
        }
        if (res.code !== 200) {
          return resolve({
            hasError: true,
            errorMessage: res.message,
            data: [],
          });
        } else {
          return resolve({
            hasError: false,
            errorMessage: null,
            data: res.layoutInfo,
          });
        }
      } catch (error) {
        return resolve({
          hasError: true,
          errorMessage: "Couldn't call text API.",
          data: [],
        });
      }
    });
    console.log("exiting backend api call");
  });
