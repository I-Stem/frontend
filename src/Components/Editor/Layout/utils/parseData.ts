import { RefObject } from "react";
import {
  createBoundingBox,
  formatLabelForAPI,
  formatLabelForCanvas,
  getCanvasLabelFromHierLabels,
  toJson,
} from "./canvasUtils";
import { BoxContent, Editor } from "../../EditorTypings";
import { BoundingBoxProperty, CanvasBoundingBoxLabel } from "../CanvasTypes";
import { saveBoundingBoxesInIndexedDB } from "./LayoutInfoDBManager";
import { getRelativeBoxesInX1Y1X2Y2FromRelativeXYWH } from "./layoutInfo";
import { getContentForBoxDigital } from "@Components/Editor/Content/ckeditor/InitialRichEditorState";
import { saveEditorStateInDB } from "@Components/Editor/Content/ConflictUtils";
import { generateHtmlForPage } from "@Components/Editor/Html/HtmlViewerHelpers";

const layoutDBName = "i-stem";

export const mapCanvasBoundingBoxesToRemediationAPISchema = (
  boundingBoxes: Array<BoundingBoxProperty>,
  zeroIndexedBasedPage: number,
  ocrJson: Editor.OCRJsonSchema | any,
  type: string,
  version: string,
  twoColumn: boolean
) => {
  const corrections: Editor.BoxCorrections[] = [];
  let layoutInfo: Editor.LayoutInfo;
  if (type === "DIGITAL") {
    layoutInfo = {
      page_width: 0,
      page_height: 0,
      page_numbers: [zeroIndexedBasedPage],
      ocr_text: ocrJson.elements[zeroIndexedBasedPage],
      table_results: [],
      corrections: [],
      line_data: "",
      word_data: "",
      detections: [],
      two_column: twoColumn,
      is_math: ocrJson.math,
      type,
      version,
    };
  } else {
    layoutInfo = {
      page_width: 0,
      page_height: 0,
      page_numbers: [zeroIndexedBasedPage],
      ocr_text: ocrJson.meta?.ocr_text
        ? ocrJson.meta?.ocr_text[zeroIndexedBasedPage]
        : ocrJson.ocr_text[zeroIndexedBasedPage],
      table_results: ocrJson.meta?.table_results
        ? ocrJson.meta?.table_results[zeroIndexedBasedPage]
        : ocrJson.table_results[zeroIndexedBasedPage] || [],
      corrections: [],
      line_data: ocrJson.meta?.line_data
        ? ocrJson.meta.line_data[zeroIndexedBasedPage]
        : "",
      word_data: ocrJson.meta?.word_data
        ? ocrJson.meta.word_data[zeroIndexedBasedPage]
        : "",
      detections: ocrJson.meta?.detections
        ? ocrJson.meta?.detections[zeroIndexedBasedPage]
        : ocrJson[zeroIndexedBasedPage].filter(
            (pageColumn: { idx: number }) => pageColumn.idx === -1
          )[0].content.detections,
      two_column: twoColumn,
      is_math: ocrJson.math,
      type,
      version,
    };
  }
  boundingBoxes.forEach(box => {
    corrections.push({
      id: box.ind,
      label: formatLabelForAPI(box.label),
      relative_box: getRelativeBoxesInX1Y1X2Y2FromRelativeXYWH([
        box.x,
        box.y,
        box.w,
        box.h,
      ]),
      reading_order: box.order,
      page: zeroIndexedBasedPage,
      is_partial: false,
      level: box.level ?? undefined,
      type: box.type ?? undefined,
      hierId: box.hierId,
      scenario: box.scenario,
      isDeleted: box.isDeleted,
      operations: box.operations,
    });
  });
  layoutInfo.corrections = corrections;
  return layoutInfo;
};
const getOCRDetectionsForPageDigital = (
  ocrJson: Editor.OCRJsonSchemaDigital,
  zeroBasedPageNumber: number
): Editor.digitalOCRElements => {
  return ocrJson.elements[zeroBasedPageNumber];
};

const getOCRDetectionsForPage = (
  ocrJson: Editor.OCRJsonSchema | any,
  zeroBasedPageNumber: number
): Editor.OCRDetections => {
  if (ocrJson.math) {
    return ocrJson.meta.detections[zeroBasedPageNumber];
  } else {
    return ocrJson[zeroBasedPageNumber].filter(
      (pageColumn: { idx: number }) => pageColumn.idx === -1
    )[0].content.detections;
  }
};

function getHierListBounds(box: Editor.HierList | any) {
  const length = box.listItems!.length;
  const firstItemBounds = box.listItems[0].Children[0].Bounds;
  let x1 = firstItemBounds[0];
  let y1 = firstItemBounds[3];
  let x2 = firstItemBounds[2];
  let y2 = firstItemBounds[1];
  box.listItems?.map((item: any) => {
    let listBox;
    if (item.marker_type == Editor.DigitalLabels.LIST) {
      listBox = getHierListBounds(item.Children[0]);
      listBox = [listBox[0], listBox[3], listBox[2], listBox[1]];
    } else listBox = item.Children[0].Bounds;
    if (listBox[0] < x1) x1 = listBox[0];
    if (listBox[3] > y1) y1 = listBox[3];
    if (x2 < listBox[2]) x2 = listBox[2];

    if (y2 > listBox[1]) y2 = listBox[1];
  });

  return [x1, y1, x2, y2];
}

function getSpanningBoxAndContent(
  boxes: any[],
  pageSize: { width: number; height: number }
) {
  let minX1 = 99999,
    minY1 = 99999;
  let maxX2 = 0,
    maxY2 = 0;
  boxes.map((child, spanningIndex) => {
    if (child.Bounds[0] < minX1) minX1 = child.Bounds[0];

    if (child.Bounds[1] < minY1) minY1 = child.Bounds[1];

    if (child.Bounds[2] > maxX2) maxX2 = child.Bounds[2];

    if (child.Bounds[3] > maxY2) maxY2 = child.Bounds[3];
  });

  return {
    bbox: [
      minX1 / pageSize.width,
      (pageSize.height - maxY2) / pageSize.height,
      maxX2 / pageSize.width,
      (pageSize.height - minY1) / pageSize.height,
    ] as Editor.BoundingBoxType,
    content: boxes.map(child => child.Text).join(" "),
  };
}

export const getBoundsForDigital = (
  box: Editor.DigitalBoxContent,
  pageSize: { height: number; width: number }
): { bbox: Editor.BoundingBoxType; content: string; scenario?: string }[] => {
  switch (box.type) {
    case Editor.DigitalLabels.HEADING:
    case Editor.DigitalLabels.PARAGRAPH:
    case Editor.DigitalLabels.DOC_TITLE:
    case Editor.DigitalLabels.FOOTNOTES:
      {
        if (box.Children.length > 1) {
          let column1 = [box.Children[0]],
            column2 = [];
          let i = 1;
          while (i < box.Children.length) {
            if (
              column1[column1.length - 1].Bounds[2] < box.Children[i].Bounds[0]
            )
              break;
            else column1.push(box.Children[i]);
            i++;
          }

          for (let j = i; j < box.Children.length; j++)
            column2.push(box.Children[j]);

          if (column2.length > 0) {
            return [
              {
                ...getSpanningBoxAndContent(column1, pageSize),
                scenario: "COLSPAN",
              },
              {
                ...getSpanningBoxAndContent(column2, pageSize),
                scenario: "COLSPAN",
              },
            ];
          } else return [getSpanningBoxAndContent(column1, pageSize)];
        } else
          return [
            {
              bbox: [
                box.Children[0].Bounds[0] / pageSize.width,
                (pageSize.height - box.Children[0].Bounds[3]) / pageSize.height,
                box.Children[0].Bounds[2] / pageSize.width,
                (pageSize.height - box.Children[0].Bounds[1]) / pageSize.height,
              ],
              content: getContentForBoxDigital(box, -1),
            },
          ];
      }
      break;

    case Editor.DigitalLabels.FIGURE:
    case Editor.DigitalLabels.TABLE:
      if (box.Bounds)
        return [
          {
            bbox: [
              box.Bounds[0] / pageSize.width,
              (pageSize.height - box.Bounds[3]) / pageSize.height,
              box.Bounds[2] / pageSize.width,
              (pageSize.height - box.Bounds[1]) / pageSize.height,
            ],
            content: getContentForBoxDigital(box, -1),
          },
        ];
      break;

    case Editor.DigitalLabels.LIST:
      const listBox = getHierListBounds(box);
      return [
        {
          bbox: [
            listBox[0] / pageSize.width,
            (pageSize.height - listBox[1]) / pageSize.height,
            listBox[2] / pageSize.width,
            (pageSize.height - listBox[3]) / pageSize.height,
          ],
          content: getContentForBoxDigital(box, -1),
        },
      ];
      break;
  }

  return [
    {
      bbox: [0, 0, 0, 0],
      content: "unknown",
    },
  ];
};

export const getOcrJsonBoundingBoxesForPage = async (
  fileId: string,
  currentPage: number,
  docType: string
): Promise<BoundingBoxProperty[]> => {
  return new Promise((resolve, reject) => {
    toJson(
      `${fileId}-json`,
      (json: Editor.OCRJsonSchema & Editor.OCRJsonSchemaDigital) => {
        const bboxes: BoundingBoxProperty[] = [];

        if (docType === "DIGITAL") {
          console.log("parsing digital document");
          const boxContents: BoxContent[] = [];
          const cropBox = json.elements[currentPage - 1].boxes.CropBox;
          const pageSize = {
            height: cropBox[3],
            width: cropBox[2],
          };

          let boxIndex = 0;
          getOCRDetectionsForPageDigital(
            json,
            currentPage - 1
          ).page_content.map((box: Editor.DigitalBoxContent) => {
            const relativeBoxes = getBoundsForDigital(box, pageSize);
            relativeBoxes.forEach(relativeBox => {
              const boxObj: BoundingBoxProperty = createBoundingBox({
                label: getCanvasLabelFromHierLabels(box.type, {
                  hLevel: box.heading_level || 1,
                }) as CanvasBoundingBoxLabel,
                x: relativeBox.bbox[0],
                y: relativeBox.bbox[1],
                w: relativeBox.bbox[2] - relativeBox.bbox[0],
                h: relativeBox.bbox[3] - relativeBox.bbox[1],
                ind: boxIndex,
                hierId: box.element_id,
                scenario: relativeBox.scenario,
                color: "#008A5AB3",
                lineWidth: 1,
                order: boxIndex + 1,
                level:
                  box.type === Editor.DigitalLabels.HEADING
                    ? CanvasBoundingBoxLabel.HEADING + box.heading_level
                    : undefined,
                isDeleted: false,
              });
              bboxes.push(boxObj);

              const boxContent: BoxContent = {
                edited: false,
                id: boxObj.ind,
                label: boxObj.label,
                boxInX1Y1X2Y2: relativeBox.bbox,
                order: boxObj.order || 1,
                editorVersion: process.env.NEXT_PUBLIC_EDITOR_VERSION,
                content: relativeBox.content,
                metadata:
                  box.type === Editor.DigitalLabels.HEADING
                    ? {
                        level:
                          CanvasBoundingBoxLabel.HEADING + box.heading_level,
                      }
                    : undefined,
              };
              boxContents.push(boxContent);
              boxIndex += 1;
            });
          });

          if (boxContents.length > 0) {
            saveEditorStateInDB(fileId, currentPage, {
              hasPreviousState: false,
              currentState: {
                state: boxContents,
                boundingBoxes: bboxes,
              },
            });
            generateHtmlForPage(fileId, currentPage, null, docType, {
              hasPreviousState: false,
              currentState: {
                state: boxContents,
                boundingBoxes: bboxes,
              },
            });
          }
        } else
          getOCRDetectionsForPage(json, currentPage - 1).boxes.map(
            (box, ind) => {
              const object: BoundingBoxProperty = createBoundingBox({
                label: formatLabelForCanvas(
                  box.label,
                  null
                ) as CanvasBoundingBoxLabel,
                x: box.relative_box[0],
                y: box.relative_box[1],
                w: box.relative_box[2] - box.relative_box[0],
                h: box.relative_box[3] - box.relative_box[1],
                ind,
                color: "#008A5AB3",
                lineWidth: 1,
                order: ind + 1,
                isDeleted: false,
              });
              bboxes.push(object);
            }
          );

        saveBoundingBoxesInIndexedDB(fileId, currentPage, bboxes);

        return resolve(bboxes);
      }
    );
  });
};
