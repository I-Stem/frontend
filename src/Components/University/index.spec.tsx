import * as React from "react";

import { render } from "@Test/utils";
// import { Rec } from "@Components";
import { RecommendedActions } from "./RecommendedActions";
import { InviteModal } from "./InviteModal";
import { MetricsReportDialog } from "./MetricsReport";
import { StudentDetails } from "./StudentDetails";
import { UserType } from "@Definitions/Constants";
const baseProps = {
  mockFunc: jest.fn(),
};
describe("Recomended Actions", () => {
  it("should render without fail", () => {
    const { getByText } = render(
      <RecommendedActions
        cardPreferences={{
          showOnboardStudentsCard: true,
          showOnboardStaffCard: true,
        }}
        updatePreferences={baseProps.mockFunc}
        userType={UserType.I_STEM}
      />
    );

    expect(getByText("RECOMMENDED ACTIONS")).toBeTruthy();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <RecommendedActions
        cardPreferences={{
          showOnboardStudentsCard: true,
          showOnboardStaffCard: true,
        }}
        updatePreferences={baseProps.mockFunc}
        userType={UserType.I_STEM}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
describe("InviteModal", () => {
  it("should render without fail", () => {
    const { getByText } = render(
      <InviteModal
        formSubmit={baseProps.mockFunc}
        isOpen={true}
        toggleModal={baseProps.mockFunc}
        invitationFor="REMEDIATORS"
      >
        Test InviteModal
      </InviteModal>
    );

    expect(getByText("Test InviteModal")).toBeTruthy();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <InviteModal
        formSubmit={baseProps.mockFunc}
        isOpen={true}
        toggleModal={baseProps.mockFunc}
        invitationFor="REMEDIATORS"
      >
        Test InviteModal
      </InviteModal>
    );

    expect(container).toMatchSnapshot();
  });
});
describe("MetricsReportDialog", () => {
  it("should render without fail", () => {
    const { getByText } = render(
      <MetricsReportDialog showModal={true} toggleDialog={baseProps.mockFunc} />
    );

    expect(
      getByText(
        "A report for students metrics will be generated as CSV and send to your email."
      )
    ).toBeTruthy();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <MetricsReportDialog showModal={true} toggleDialog={baseProps.mockFunc} />
    );

    expect(container).toMatchSnapshot();
  });
});
describe("StudentDetails", () => {
  it("should render without fail", () => {
    const { getByRole } = render(
      <StudentDetails
        studentId={(undefined as unknown) as string}
        studentDetails={{
          name: "ABCD",
          email: "abc@example.com",
          roll: "1234",
        }}
      />
    );

    expect(getByRole("heading")).toBeTruthy();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <StudentDetails
        studentId={(undefined as unknown) as string}
        studentDetails={{
          name: "ABCD",
          email: "abc@example.com",
          roll: "1234",
        }}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
