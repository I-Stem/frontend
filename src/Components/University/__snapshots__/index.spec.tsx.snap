// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recomended Actions should match snapshot 1`] = `
<div>
  <div>
    <h2
      class="font-semibold text-xl heading-color"
    >
      RECOMMENDED ACTIONS
    </h2>
    <div
      class="onboard-tiles row"
    >
      <div
        class="col-sm-6"
        style="margin-top: 1rem;"
      >
        <h3
          class="font-semibold text-l heading-color"
        >
          Get started by inviting your students to the I-Stem platform.
        </h3>
        <p>
          Or you can always go to Students &gt; INVITE STUDENTS / IMPORT STUDENTS to invite students.
        </p>
        <div
          class="onboard-button"
        >
          <div
            class="w-full"
          >
            <button
              class="ant-btn ant-btn-block text-center blue-btn"
              type="button"
            >
              <span
                class="font-semibold text-lg text-white leading-6"
              >
                ONBOARD STUDENTS
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="col-sm-4"
        style="margin-left: 8rem;"
      >
        <img
          alt="decorative"
          aria-hidden="true"
          class="w-full lip-img"
          src="/static/images/onboard-students.svg"
        />
      </div>
      <div>
        <img
          alt="Close tile"
          class="close-tiles"
          role="button"
          src="/static/images/closeIcon.svg"
        />
      </div>
    </div>
    <div
      class="onboard-tiles row"
    >
      <div
        class="col-sm-6"
        style="margin-top: 1rem;"
      >
        <h3
          class="font-semibold text-l heading-color"
        >
          Get started by inviting your staff to the I-Stem platform.
        </h3>
        <p>
          Or you can always go to Settings &gt; INVITE STAFF to invite staff.
        </p>
        <div
          class="onboard-button"
        >
          <div
            class="w-full"
          >
            <button
              class="ant-btn ant-btn-block text-center blue-btn"
              type="button"
            >
              <span
                class="font-semibold text-lg text-white leading-6"
              >
                ONBOARD STAFF
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="col-sm-4"
        style="margin-left: 8rem;"
      >
        <img
          alt="decorative"
          aria-hidden="true"
          class="w-full lip-img"
          src="/static/images/onboard-staff.svg"
        />
      </div>
      <div>
        <img
          alt="Close tile"
          class="close-tiles"
          role="button"
          src="/static/images/closeIcon.svg"
        />
      </div>
    </div>
  </div>
</div>
`;
